security/l1-contracts-dev-flow-vision.md

```diff
@@ -0,0 +1,159 @@
+
+For example, if you are adding something to the SystemConfig that will ship in the next release, make
+those changes directly to `SystemConfig.sol`. Tests would be added directly to `SystemConfig.t.sol` like normal.
+
+If your feature may not go into the next release, create a child contract such as `contract SystemConfigInterop is SystemConfig`.
+In this contract you can add your feature, and once we know it's ready for release it will be merged
+into the parent contract it inherits from. Tests for `SystemConfigInterop` would live in `SystemConfigInterop.t.sol`
+(TODO flesh out more details on how this should work with OPCM, e.g. do we need an `OPCM*` for each feature, like
+`OPCMInterop`, etc? A simpler approach may be to always use OPCM as the base standard deploy, and write a `interopUpgrade()`
+method that contains the code that will eventually go in `OPCM.upgrade`).
+
+An alternate approach to inheritance is to put code directly into `SystemConfig`, but behind a
+feature flag. This is the approach client code uses so they can ship releases without also shipping
+features that are not production ready. However, in safety-critical systems dead code is typically
+not allowed, so we likely will not allowed this approach for contracts. However, we mention it here
+anyway for completeness.
+
+## Testing
+
+In the above section we described how standard unit tests would be written during development. Another
+aspect is how deploy and upgrade scripts are written and tested during development, as this is a key
+change.
+
+Originally, `Deploy.s.sol` was the script that run as part of test `setUp`, so we can run tests
+against the state resulting from the deploy script. In practice, that `Deploy.s.sol` script was
+never used for real deploys. So `Deploy.s.sol` has been refactored to instead use OPCM to deploy
+a chain's L1 contracts and use that fresh deploy as the base for tests. This means that all tests
+will be known to pass against the state resulting after a new chain deploy. (In the future, we may
+delete the full `Deploy.s.sol` file and replace it with OPCM code directly, as embedding OPCM into
+this legacy deploy script was only done as it's a simpler transition).
+
+OPCM contains a `deploy()` method that takes some arguments to deploy a chain. Some changes to
+source code will require no changes to this method—for example, if you only change things like
+code comments or function internals. Other changes will require changes to this method, like
+adding new contracts or modifying the `initialize` signature. For the latter case, because OPCM
+is what's used to scaffold chains for testing, you will be forced to modify the `deploy` method to
+account for your changes in order to test them. This is great, as it's now trivial to keep deploy
+scripts up to date, and we essentially get this for free.
+
+OPCM will also contain an `upgrade` method as described in the [OPCM Upgrades design doc](l1-upgrades.md).
```

Comment:
Just to be super clear, the `upgrade` method is used to upgrade from the previous version to the same version that is deployed via the `deploy` method

Reply:
Correct.